dlib/CMakeLists.txt

```diff
@@ -47,7 +47,15 @@ endif()
 
 # default to a Release build (except if CMAKE_BUILD_TYPE is set)
 include(cmake_utils/release_build_by_default)
-include(cmake_utils/use_cpp_11.cmake)
+
+function(enable_cpp14_for_target target_name)
+   set_target_properties(${target_name} PROPERTIES
+      CXX_STANDARD 17
```

Comment:
Shouldn't this be 14?

Reply:
Yes you're right sorry. But this totally broke everything. So I might wait for David to commit all the 14 stuff then I'll merge

Reply:
I do like cmake but i can understand why modern languages laugh at the C++ community because of our rubbish build tools. I wonder if meson or other ones are actually better, and if so, if they should get standardised just so that everyone is pushed to use to the same thing.

Reply:
In my opinion, build tools and documentation are more important than the language itself, and the standards community should be putting 100% of their efforts towards improving build time, build tools and inline documentation

Reply:
Na it's not cmake.  It's easy to have real clean cmake. It's all complicated because there are so many people with broken systems or compilers or whatever.  You can make it so simple.  But then someone posts "hey, I'm building on $some_weird_linux_distro_or_whatever and I'm getting this linker error saying libpng is missing some symbol".  Or they are complaining about something else that is actually broken on their system.  For instance, Anaconda includes these copies of x11 header files right, which for a lot of people end up in their compiler's include search path.  And Anaconda has helpfully edited those headers to be different from the real x11 headers, so if those get included instead the world breaks.   I could go on and on.

Anyway, those people complain to me though.  "Hey, dlib isn't building?!".  So I add some logic to the dlib cmake scripts to detect whatever is wrong with their system and work around it.  Like all that logic in there for testing if libpng is *really* there or not.

The only way this whole mess could be resolved was if applications *never ever* depended on libraries they are not themselves building.  And the building of those libraries would have to be controlled by one unified build system.  But that's a tall ask because everyone then turns around and says "hey, I hate the build times.  Can't the $(os, package system, whatever) provide precompiled binaries instead?" which is a desire 100% contradictory to resolving this problem given the nature of the One Definition Rule and the diversity of compile and build settings.

Then you top this all off with proprietary vendor libraries.  Like IntelMKL or anything CUDA related.  Which you can't build.  They have to be somehow pulled in as binaries because that's the only things those vendors will give out.  And those binaries have OS dependencies and even driver level dependencies. 

So IDK.  Any language that wants to do anything with hardware acceleration has to grapple with this.  I guess in the dream world there would be BLAS, LAPACK, and CUDA built into compilers by default.  Which is just saying that we would foist the problem of that stuff onto compiler and OS vendors and they would just deal with it.

Reply:
Anyway, let me see about adjusting cmake settings to use C++14.  I'm doing that right now :)  I'm going to burn away some of the oldest cmake logic for old broken compilers too and see who complains.

Reply:
> Na it's not cmake. It's easy to have real clean cmake. It's all complicated because there are so many people with broken systems or compilers or whatever. You can make it so simple. But then someone posts "hey, I'm building on $some_weird_linux_distro_or_whatever and I'm getting this linker error saying libpng is missing some symbol". Or they are complaining about something else that is actually broken on their system. For instance, Anaconda includes these copies of x11 header files right, which for a lot of people end up in their compiler's include search path. And Anaconda has helpfully edited those headers to be different from the real x11 headers, so if those get included instead the world breaks. I could go on and on.
> 
> Anyway, those people complain to me though. "Hey, dlib isn't building?!". So I add some logic to the dlib cmake scripts to detect whatever is wrong with their system and work around it. Like all that logic in there for testing if libpng is _really_ there or not.
> 
> The only way this whole mess could be resolved was if applications _never ever_ depended on libraries they are not themselves building. And the building of those libraries would have to be controlled by one unified build system. But that's a tall ask because everyone then turns around and says "hey, I hate the build times. Can't the $(os, package system, whatever) provide precompiled binaries instead?" which is a desire 100% contradictory to resolving this problem given the nature of the One Definition Rule and the diversity of compile and build settings.
> 
> Then you top this all off with proprietary vendor libraries. Like IntelMKL or anything CUDA related. Which you can't build. They have to be somehow pulled in as binaries because that's the only things those vendors will give out. And those binaries have OS dependencies and even driver level dependencies.
> 
> So IDK. Any language that wants to do anything with hardware acceleration has to grapple with this. I guess in the dream world there would be BLAS, LAPACK, and CUDA built into compilers by default. Which is just saying that we would foist the problem of that stuff onto compiler and OS vendors and they would just deal with it.

This just illustrates something is wrong with the language I think.